### PR TITLE
chore(codeintel): Differentiate between paths relative to upload root vs repo root

### PR DESCRIPTION
**Pull request by varungandhi-src**
Should hopefully make the code easier to follow due to the mixture of
different kinds of paths in different places.

## Test plan

TODO:

- [x] Update existing tests
- [x] Locally upload scip-go index at different roots for sg monorepo and check that
       codenav is working OK - checked for `lib` subdir

## Changelog

### CODE REVIEW

File: internal/codeintel/codenav/gittree_translator_test.go [hunk 1]
```diff
@@ -23,7 +24,7 @@ func TestGetTargetCommitPathFromSourcePath(t *testing.T) {
 	args := &requestArgs{
 		repo:   &sgtypes.Repo{ID: 50},
 		commit: "deadbeef1",
-		path:   "/foo/bar.go",
+		path:   core.NewRepoRootRelativePathUnchecked("/foo/bar.go"),
```
Comment (kritzcreek) [inline internal/codeintel/codenav/gittree_translator_test.go:27]: Why are we making unchecked relative paths from absolute paths here?

File: internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go [hunk 2]
```diff
@@ -4,6 +4,7 @@ import (
 	"bytes"
 	"context"
 	"fmt"
+	"github.com/sourcegraph/sourcegraph/internal/codeintel/core"
```
Comment (kritzcreek) [inline internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go:7]: import formatting here, and in a couple places. (Can we really not get `go fmt` to do this?)

File: internal/codeintel/codenav/transport/graphql/root_resolver_test.go [hunk 3]
```diff
@@ -29,13 +30,15 @@ import (
...
+var repoRootPath = core.NewRepoRootRelativePathUnchecked
+
 func TestRanges(t *testing.T) {
 	mockCodeNavService := NewMockCodeNavService()
 
 	mockRequestState := codenav.RequestState{
 		RepositoryID: 1,
 		Commit:       "deadbeef1",
-		Path:         "/src/main",
+		Path:         repoRootPath("/src/main"),
```
Comment (kritzcreek) [inline internal/codeintel/codenav/transport/graphql/root_resolver_test.go:41]: These also seem to all be absolute paths.

File: internal/codeintel/core/paths.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+type RepoRootRelativePath struct {
+	rawValue string
+}
+
+func NewRepoRootRelativePathUnchecked(s string) RepoRootRelativePath {
+	return RepoRootRelativePath{rawValue: s}
+}
+
+// This function takes a string as the first argument instead of a *shared.CompletedUpload
+// to avoid a dependency on a higher-level package
```
Comment (kritzcreek) [inline internal/codeintel/core/paths.go:41]: Can't you avoid the dependency using interfaces? (That's the one thing Go's interfaces are good at after all...)

**Review by kritzcreek — CHANGES_REQUESTED** [review]
I think it's a good idea to un-stringify these.

Naming is always hard, but maybe we can find something a little shorter to reduce the clutter? How about `RepoRelPath` & `UploadRelPath`?

File: internal/codeintel/codenav/gittree_translator_test.go (hunk 1 above)
Comment (varungandhi-src) [reply internal/codeintel/codenav/gittree_translator_test.go:27]: It's like the weird git thing where gitignore uses "absolute paths" from the repo root. I can remove the extra leading `/`.

File: internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go (hunk 2 above)
Comment (varungandhi-src) [reply internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go:7]: There is a separate tool goimports tool, I just configured GoLand to do this. https://docs-legacy.sourcegraph.com/dev/background-information/goland#imports-style

There was some discussion around enabling this generally. https://sourcegraph.slack.com/archives/C04MYFW01NV/p1713164404103499

File: internal/codeintel/core/paths.go (hunk 4 above)
Comment (varungandhi-src) [reply internal/codeintel/core/paths.go:41]: That's a good point, I didn't consider it earlier.

Interfaces can only have methods, not fields. So introducing a single interface method for a single field would be a bit weird (and it's not terribly common). But I'll consider doing this. Leaving this open for now.

File: internal/codeintel/codenav/gittree_translator_test.go (hunk 1 above)
Comment (kritzcreek) [reply internal/codeintel/codenav/gittree_translator_test.go:27]: Oh I see, interesting 🤔 Is that a path scheme we're using in other places? For example right now an upload at the root path of a repository will use `""` as its path in the database.

File: internal/codeintel/core/paths.go (hunk 4 above)
Comment (kritzcreek) [reply internal/codeintel/core/paths.go:41]: FWIW I did the same for the occurrence matching code https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph@b47c376cbef0798106eb5a849fc6786ad37c8fe7/-/blob/internal/codeintel/codenav/scip_utils.go?L8-10

File: internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go (hunk 2 above)
Comment (kritzcreek) [reply internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go:7]: Thanks for the pointers. I've set up Goland with those settings, let's see if it helps :)

File: internal/codeintel/codenav/gittree_translator_test.go (hunk 1 above)
Comment (varungandhi-src) [reply internal/codeintel/codenav/gittree_translator_test.go]: I think it's only used in these tests, even the other tests don't use it.

Comment (varungandhi-src) [reply internal/codeintel/codenav/gittree_translator_test.go]: Removed that. It's a bit unfortunate, changing this is going to cause merge conflicts with my other PR, but that's OK, I guess...

File: internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go (hunk 2 above)
Comment (varungandhi-src) [reply internal/codeintel/codenav/internal/lsifstore/locations_by_position_test.go]: Fixed the import ordering.

File: internal/codeintel/core/paths.go (hunk 4 above)
Comment (varungandhi-src) [reply internal/codeintel/core/paths.go:51]: Go has enough odd naming conventions, let's avoid using naming conventions strongly associated with other languages, if only to have people on other teams feel comfortable when reading our code? 😅 In this case, I've created an interface called `UploadLike`, happy to rename that to match our existing conventions if you have a better idea.

Comment (kritzcreek) [reply internal/codeintel/core/paths.go:51]: `UploadLike` works for me. Please also update the comments on `NewUploadRelPath` and `NewRepoRelPath`

**Review by kritzcreek — APPROVED** [review]

**Pull request merged**